homeassistant/components/camera/xeoma.py

```diff
@@ -0,0 +1,112 @@
+"""
+Support for Xeoma Cameras.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/camera.xeoma/
+"""
+import asyncio
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.camera import (
+    PLATFORM_SCHEMA, Camera)
+from homeassistant.const import (
+    CONF_NAME, CONF_USERNAME, CONF_PASSWORD, CONF_HOST)
+from homeassistant.helpers import config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+REQUIREMENTS = ['pyxeoma==1.2']
+
+CONF_CAMERAS = 'cameras'
+CONF_HIDE = 'hide'
+CONF_IMAGE_NAME = 'image_name'
+CONF_NEW_VERSION = 'new_version'
+
+CAMERAS_SCHEMA = vol.Schema({
+    vol.Required(CONF_IMAGE_NAME): cv.string,
+    vol.Optional(CONF_HIDE, default=False): cv.boolean,
+    vol.Optional(CONF_NAME): cv.string
+}, required=False)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_CAMERAS, default={}):
+        vol.Schema(vol.All(cv.ensure_list, [CAMERAS_SCHEMA])),
+    vol.Optional(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_USERNAME): cv.string,
+    vol.Optional(CONF_NEW_VERSION, default=True): cv.boolean
+})
+
+
+@asyncio.coroutine
+# pylint: disable=unused-argument
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Discover and setup Xeoma Cameras."""
+    host = config[CONF_HOST]
+    login = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    new_version = config[CONF_NEW_VERSION]
+
+    from pyxeoma.xeoma import Xeoma, XeomaError
+    xeoma = Xeoma(host, new_version, login, password)
```

Comment:
I assume that there will be a trackback if the credentials are wrong or the camera is not reachable. Please catch that,  log an error that the users knows what's going on and return. Otherwise the users end up with a non-functional platform.

Reply:
Xeoma is an NVR similar to BlueIris or Zoneminder so the connection being made is not to individual cameras, but to a web interface that can be enabled within Xeoma which provides direct access to jpgs from all connected cameras. The constructor for the Xeoma object doesn't make any attempt to connect to the Xeoma web server. All of the calls to the library that can fail are inside the following `try` block. The first function call inside the `try` block, `xeoma.async_test_connection()`, will test the connection to the server. If this fails, the library raises a `XeomaError` with an appropriate message and the platform logs that.

As far as authentication, there are 2 possible failure scenarios that I can see:

1. Incorrect credentials are provided: The library will attempt to obtain a session key, but fail and it will raise a `XeomaError`. 
2. No credential are provided, but credentials are required: The library will attempt to discover cameras, but none will be found and it will raise a `XeomaError`.

In both cases the platform will catch and log the error and since it will only add entities to home assistant for cameras it is able to discover, no non-working camera entities will be added.

The main issue is that the web server that Xeoma provides doesn't handle HTTP status codes correctly. In scenario 2 above, you still get a 200 response which makes it difficult to determine exactly why camera discovery failed since incorrect configuration of the web server module within Xeoma can also create a situation in which no cameras will be discovered (though this should be obvious since the user won't be able to access their cameras in the Xeoma web ui either). The error message provided in this scenario simply informs the user that no cameras were discovered.

Reply:
`pyxeoma` is handling all connection issues. So it makes sense to just push then through.